File: server/query_syntax.py
"""The typed search syntax: `user:alice type:add status:failed some free text`.

One search box, two jobs: structured filters for what you know, and free text
for what you half-remember. Splitting them into separate inputs would be more
honest to the implementation and worse to use, so they share a box and are
separated here.

Unknown prefixes are deliberately *not* treated as filters. `http://example.com`
would otherwise parse as `http:` with a value, and a typo like `usr:alice` would
silently filter on nothing and return everything. Both fall through to free text,
where they at least behave like a search.

Mirrored in dashboard/src/lib/query-syntax.ts, which parses the same grammar to
render filter chips before the request is sent.
"""

from __future__ import annotations

import re
import shlex
from dataclasses import dataclass, field
from typing import Optional

# The entity prefixes, mapped to the entity type they filter on. `session` is
# the UI's word for what the API calls a run; both are accepted.
ENTITY_PREFIXES: dict[str, str] = {
    "user": "user",
    "agent": "agent",
    "run": "run",
    "session": "run",
    "app": "agent",
}

TYPE_VALUES = {"add", "search", "get_all", "get", "update", "delete", "delete_all", "other"}
STATUS_VALUES = {"succeeded", "failed"}
METHOD_VALUES = {"GET", "POST", "PUT", "PATCH", "DELETE"}

# Aliases people actually type for request types.
TYPE_ALIASES = {
    "getall": "get_all",
    "get-all": "get_all",
    "list": "get_all",
    "deleteall": "delete_all",
    "delete-all": "delete_all",
    "create": "add",
    "write": "add",
    "query": "search",
}

_TOKEN = re.compile(r"^(?P<key>[a-zA-Z_-]+):(?P<value>.*)$")


@dataclass(frozen=True)
class ParsedQuery:
    types: list[str] = field(default_factory=list)
    status: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    method: Optional[str] = None
    text: Optional[str] = None

    @property
    def is_empty(self) -> bool:
        return not any((self.types, self.status, self.entity_type, self.entity_id, self.method, self.text))


EMPTY = ParsedQuery()


def _tokenize(raw: str) -> list[str]:
    """Split on whitespace, honouring quotes so `user:"ada lovelace"` survives."""
    try:
        return shlex.split(raw)
    except ValueError:
        # Unbalanced quote - someone is mid-typing. Fall back to a plain split
        # rather than refusing to search.
        return raw.split()


def parse(raw: Optional[str]) -> ParsedQuery:
    if not raw or not raw.strip():
        return EMPTY

    types: list[str] = []
    status: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    method: Optional[str] = None
    free: list[str] = []

    for token in _tokenize(raw):
        match = _TOKEN.match(token)
        if not match:
            free.append(token)
            continue

        key = match.group("key").lower()
        value = match.group("value").strip()
        if not value:
            free.append(token)
            continue

        if key in ENTITY_PREFIXES:
            entity_type = ENTITY_PREFIXES[key]
            entity_id = value
        elif key == "type":
            normalized = TYPE_ALIASES.get(value.lower(), value.lower())
            if normalized in TYPE_VALUES:
                types.append(normalized)
            else:
                free.append(token)
        elif key == "status":
            lowered = value.lower()
            if lowered in STATUS_VALUES:
                status = lowered
            elif lowered in ("ok", "success", "2xx"):
                status = "succeeded"
            elif lowered in ("error", "fail", "4xx", "5xx"):
                status = "failed"
            else:
                free.append(token)
        elif key == "method":
            upper = value.upper()
            if upper in METHOD_VALUES:
                method = upper
            else:
                free.append(token)
        else:
            # Unknown prefix. Not a filter - see the module docstring.
            free.append(token)

    return ParsedQuery(
        types=types,
        status=status,
        entity_type=entity_type,
        entity_id=entity_id,
        method=method,
        text=" ".join(free) or None,
    )

File: server/dashboard/src/lib/query-syntax.ts
/**
 * The typed search syntax, parsed client-side so filter chips can render
 * before the request is sent.
 *
 * Mirrors server/query_syntax.py. The server is the authority — it parses the
 * raw string again and never trusts what the client derived — but the box has
 * to show what it understood as you type, and a round trip per keystroke to
 * find that out would be its own problem.
 *
 * Unknown prefixes deliberately fall through to free text. `usr:alice` filtering
 * on nothing would silently return everything, and `https://x` would parse as a
 * `https:` filter.
 */

export const ENTITY_PREFIXES: Record<string, "user" | "agent" | "run"> = {
  user: "user",
  agent: "agent",
  run: "run",
  // The UI says Sessions and Apps where the API says run_id and agent_id.
  session: "run",
  app: "agent",
};

export const TYPE_VALUES = [
  "add",
  "search",
  "get_all",
  "get",
  "update",
  "delete",
  "delete_all",
  "other",
] as const;

export type RequestType = (typeof TYPE_VALUES)[number];

const TYPE_ALIASES: Record<string, RequestType> = {
  getall: "get_all",
  "get-all": "get_all",
  list: "get_all",
  deleteall: "delete_all",
  "delete-all": "delete_all",
  create: "add",
  write: "add",
  query: "search",
};

const STATUS_SYNONYMS: Record<string, "succeeded" | "failed"> = {
  succeeded: "succeeded",
  ok: "succeeded",
  success: "succeeded",
  "2xx": "succeeded",
  failed: "failed",
  error: "failed",
  fail: "failed",
  "4xx": "failed",
  "5xx": "failed",
};

const METHODS = new Set(["GET", "POST", "PUT", "PATCH", "DELETE"]);

export interface ParsedQuery {
  types: RequestType[];
  status?: "succeeded" | "failed";
  entityType?: "user" | "agent" | "run";
  entityId?: string;
  method?: string;
  text?: string;
}

export const EMPTY_QUERY: ParsedQuery = { types: [] };

/** Split on whitespace, honouring quotes so `user:"ada lovelace"` survives. */
function tokenize(raw: string): string[] {
  const tokens: string[] = [];
  let current = "";
  let quote: string | null = null;

  for (const char of raw) {
    if (quote) {
      if (char === quote) quote = null;
      else current += char;
      continue;
    }
    if (char === '"' || char === "'") {
      quote = char;
      continue;
    }
    if (/\s/.test(char)) {
      if (current) tokens.push(current);
      current = "";
      continue;
    }
    current += char;
  }
  if (current) tokens.push(current);
  return tokens;
}

export function parseQuery(raw: string | undefined | null): ParsedQuery {
  if (!raw || !raw.trim()) return EMPTY_QUERY;

  const parsed: ParsedQuery = { types: [] };
  const free: string[] = [];

  for (const token of tokenize(raw)) {
    const match = /^([a-zA-Z_-]+):(.*)$/.exec(token);
    const value = match?.[2]?.trim();
    if (!match || !value) {
      free.push(token);
      continue;
    }
    const key = match[1].toLowerCase();

    if (key in ENTITY_PREFIXES) {
      parsed.entityType = ENTITY_PREFIXES[key];
      parsed.entityId = value;
    } else if (key === "type") {
      const normalized = (TYPE_ALIASES[value.toLowerCase()] ??
        value.toLowerCase()) as RequestType;
      if ((TYPE_VALUES as readonly string[]).includes(normalized)) {
        parsed.types.push(normalized);
      } else {
        free.push(token);
      }
    } else if (key === "status") {
      const normalized = STATUS_SYNONYMS[value.toLowerCase()];
      if (normalized) parsed.status = normalized;
      else free.push(token);
    } else if (key === "method") {
      const upper = value.toUpperCase();
      if (METHODS.has(upper)) parsed.method = upper;
      else free.push(token);
    } else {
      free.push(token);
    }
  }

  if (free.length) parsed.text = free.join(" ");
  return parsed;
}

/** Human-readable chips for what the query resolved to. */
export function describeQuery(parsed: ParsedQuery): string[] {
  const chips: string[] = [];
  for (const type of parsed.types) chips.push(`type: ${type}`);
  if (parsed.status) chips.push(`status: ${parsed.status}`);
  if (parsed.entityId) chips.push(`${parsed.entityType}: ${parsed.entityId}`);
  if (parsed.method) chips.push(`method: ${parsed.method}`);
  if (parsed.text) chips.push(`text: ${parsed.text}`);
  return chips;
}

export const QUERY_HINTS = [
  "user:alice",
  "agent:aurion",
  "session:s-42",
  "type:add",
  "type:search",
  "status:failed",
  "method:post",
];
